src/keyboard.cpp

```diff
@@ -891,308 +724,520 @@ int kbd_layout[4][KBD_MAX_ENTRIES][2] = {
+	{ "SDLK_F15", SDLK_F15},
+	/*@}*/
+		/** @name Key state modifier keys */
+	/*@{*/
+	{ "SDLK_NUMLOCK", SDLK_NUMLOCK},
+	{ "SDLK_CAPSLOCK", SDLK_CAPSLOCK},
+	{ "SDLK_SCROLLOCK", SDLK_SCROLLOCK},
+	{ "SDLK_RSHIFT", SDLK_RSHIFT},
+	{ "SDLK_LSHIFT", SDLK_LSHIFT},
+	{ "SDLK_RCTRL", SDLK_RCTRL},
+	{ "SDLK_LCTRL", SDLK_LCTRL},
+	{ "SDLK_RALT", SDLK_RALT},
+	{ "SDLK_LALT", SDLK_LALT},
+	{ "SDLK_RMETA", SDLK_RMETA},
+	{ "SDLK_LMETA", SDLK_LMETA},
+	{ "SDLK_LSUPER", SDLK_LSUPER},		/**< Left "Windows" key */
+	{ "SDLK_RSUPER", SDLK_RSUPER},		/**< Right "Windows" key */
+	{ "SDLK_MODE", SDLK_MODE},		/**< "Alt Gr" key */
+	{ "SDLK_COMPOSE", SDLK_COMPOSE},		/**< Multi-key compose key */
+	/*@}*/
+		/** @name Miscellaneous function keys */
+	/*@{*/
+	{ "SDLK_HELP", SDLK_HELP},
+	{ "SDLK_PRINT", SDLK_PRINT},
+	{ "SDLK_SYSREQ", SDLK_SYSREQ},
+	{ "SDLK_BREAK", SDLK_BREAK},
+	{ "SDLK_MENU", SDLK_MENU},
+	{ "SDLK_POWER", SDLK_POWER},		/**< Power Macintosh power key */
+	{ "SDLK_EURO", SDLK_EURO},		/**< Some european keyboards */
+	{ "SDLK_UNDO", SDLK_UNDO},		/**< Atari keyboard has Undo */
+       /*@}*/
+	{ "MOD_PC_SHIFT", MOD_PC_SHIFT},
+	{ "MOD_PC_CTRL", MOD_PC_CTRL},
+	{ "MOD_PC_MODE", MOD_PC_MODE},
+	{ "MOD_PC_META", MOD_PC_META},
+	{ "MOD_PC_ALT", MOD_PC_ALT}
 };
+
+int kbd_layout[KBD_MAX_ENTRIES][2];
+std::map<char, std::pair<SDLKey, SDLMod>> SDLkeysFromChars;
+
+void create_SDL_keymap(void)
+{
+	CPC_KEYS cpc_key;
+	unsigned int sdl_moddedkey;
+
+	for (auto const &item : CPCkeysFromChars) {
+		sdl_moddedkey = 0xffff;
+		cpc_key = item.second;
+		// TODO (sebhz) we should really replace all this by a map !
+		for (int i=0; i < KBD_MAX_ENTRIES; i++) {
+			if (kbd_layout[i][0] == cpc_key) {
+				sdl_moddedkey = kbd_layout[i][1];
+				break;
+			}
+		}
+		if (sdl_moddedkey != 0xffff)
+			SDLkeysFromChars[item.first] = std::make_pair(static_cast<SDLKey>(sdl_moddedkey & 0xffff), static_cast<SDLMod>(sdl_moddedkey >> 16));
+	}
+}
+
+// Format of a line: CPC_xxx\tSDLK_xxx\tMODIFIER
+// Last field is optional
+bool parse_line(char *s, unsigned int line)
+{
+		unsigned int keyv = 0;
+		
+		char *pch = strtok(s, "\t");
+		if (pch == nullptr || pch[0] == '#')
+			return false;
+
+		if (CPCkeysFromStrings.count(pch) == 0) {
+			LOG_ERROR("Unknown CPC key " << pch << " found in mapping file. Ignoring it.");
+			return false;
+		}
+		
+		for (unsigned int field=0; field < 3; field++) {
+			switch (field) {
+				case 0:
+					kbd_layout[line][0] = CPCkeysFromStrings[pch];
+					break;
+				case 1:
+				case 2:
+					if (SDLkeysFromStrings.count(pch) == 0) {
+						LOG_ERROR("Unknown SDL key or modifier " << pch << " found in mapping file. Ignoring it.");
+						return false;
+					}
+					keyv |= SDLkeysFromStrings[pch];
+					break;
+				default:
+					break;
+			}
+			pch = strtok(nullptr, "\t");
+			if (pch == nullptr)
+				break;	
+		}
+		kbd_layout[line][1] = keyv;
+		return true;
+}
+
+// TODO (sebhz) there must be a way to to this more efficiently !
```

Comment:
I simply meant that there is probably a way to do a simple pointer assignment for this, rather than iterating over the table to to a copy of a constnt table, but C++ seems to be pretty strict about this. Anyway, if we end up using maps all over the board this will eventually disappear. So I removed the comment also.